include/superior_mysqlpp/extras/prepared_statement_utils.hpp

```diff
@@ -11,89 +11,127 @@ namespace SuperiorMySqlpp
+            }
+
+            std::tuple<Args&...>(values...) = ps.getResult();
         }
+    }
 
-        /**
-         * @brief Makes prepared statement by deducing ResultBindings template arguments from plain old C function's signature
-         * @param connection Connection handle into database
-         * @param query Query for prepared statement
-         * @tparam storeResult Boolean indicating if results will be in `store` or `use` mode
-         * @tparam validateMode Indicates validate mode level
-         * @tparam warnMode Indicates warning mode level
-         * @tparam ignoreNullable Disables null type checking
-         */
-        template <bool storeResult, ValidateMetadataMode validateMode, ValidateMetadataMode warnMode, bool ignoreNullable, typename... Args>
-        auto generatePreparedStatementImpl(Connection &connection, const std::string &query, void (*)(Args...)) {
-            return connection.makePreparedStatement<ResultBindings<std::remove_cv_t<std::remove_reference_t<Args>>...>, storeResult, validateMode, warnMode, ignoreNullable>(query);
+    struct EmptyParamCallback
+    {
+        inline constexpr bool operator()()
+        {
+            return false;
         }
+    };
+
+    /**
+     * @brief Builds prepared statement from query string and constructs prepared stament with updatable arguments
+     * @param query Query to be executed (in most cases there will be selections)
+     * @param connection Connection handle into database
+     * @param paramsSetter function setting statement's parameters, returning bool if input data are available
+     * @param processingFunction function to be invoked on every row
+     *                           Its parameters must correspond with result columns (their types and count)
+     * @tparam storeResult Boolean indicating if results will be in `store` or `use` mode
+     * @tparam validateMode Indicates validate mode level
+     * @tparam warnMode Indicates warning mode level
+     * @tparam ignoreNullable Disables null type checking
+     */
+    template<bool storeResult=detail::PreparedStatementsDefault::getStoreResult(),
+             ValidateMetadataMode validateMode=detail::PreparedStatementsDefault::getValidateMode(),
+             ValidateMetadataMode warnMode=detail::PreparedStatementsDefault::getWarnMode(),
+             bool ignoreNullable=detail::PreparedStatementsDefault::getIgnoreNullable(),
+             typename ResultCallable,
+             typename ParamCallable,
+             typename ConnType>
+    void psQuery(const std::string &query, ConnType &&connection, ParamCallable &&paramsSetter, ResultCallable &&processingFunction)
+    {
+        static constexpr bool NoParamCallback = std::is_same<ParamCallable, EmptyParamCallback>::value;
+        static constexpr bool NoResultCallback = std::is_same<ResultCallable, EmptyParamCallback>::value;
```

Comment:
I called for renaming EmptyCallback to EmptyParamCallback precisely because I didn't expect you to use it for ResultCallback, as their types are logically distinct - EmptyParamCallback returns boolean, EmptyResultCallback does nothing.
That aside, I belive this function will be nicer if you negate the conditions, i.e. call it `hasParamCallback`.
Also, unless you're following some specific formatting style, I wound't use brackets around `break`.